hudi-common/src/main/java/org/apache/hudi/keygen/constant/KeyGeneratorOptions.java

```diff
@@ -34,19 +35,30 @@
     description = "")
 public class KeyGeneratorOptions extends HoodieConfig {
 
-  public static final ConfigProperty<String> URL_ENCODE_PARTITIONING = ConfigProperty
-      .key("hoodie.datasource.write.partitionpath.urlencode")
-      .defaultValue("false")
-      .markAdvanced()
-      .withDocumentation("Should we url encode the partition path value, before creating the folder structure.");
-
   public static final ConfigProperty<String> HIVE_STYLE_PARTITIONING_ENABLE = ConfigProperty
       .key("hoodie.datasource.write.hive_style_partitioning")
       .defaultValue("false")
       .withDocumentation("Flag to indicate whether to use Hive style partitioning.\n"
           + "If set true, the names of partition folders follow <partition_column_name>=<partition_value> format.\n"
           + "By default false (the names of partition folders are only partition values)");
 
+  public static final ConfigProperty<String> URL_ENCODE_PARTITIONING = ConfigProperty
+      .key("hoodie.datasource.write.partitionpath.urlencode")
+      .defaultValue("false")
+      .markAdvanced()
+      // If hive style partitioning is enabled, then url encoding is also enabled.
+      // We need to do so otherwise the partition structure is awkward in some cases.
```

Comment:
Can hive query be compatible with this change? Can hive query a partition encoded like this `partition=dd%2Fmm%2Fyy`?

Reply:
Indeed, this is a breaking change. For meta sync, we will need users to drop the table and resync.

Reply:
yeah, can we just resolve the issue by detecting the slaces `/` then encode the URL path, instead of just change all the table paths.

And we should introduce a new option to control this option so that when user encounter regression, they have an option to rollback.

Reply:
I agree with Danny that we should not change the default here.  For such cases where `partition=yyyy/mm/dd` is generated, we should throw a warning to let user know (e.g., use `yyyy-mm-dd` instead of `yyyy/mm/dd`).  In our test, we should avoid such patterns.